### PR TITLE
Gdb server #2168

### PR DESCRIPTION
**Pull request by kumargu**
## Reason for This PR

Initial implementation of a GDB server for guest debugging (https://github.com/firecracker-microvm/firecracker/pull/2168)

## Description of Changes

- Created a new package implementing the gdbstub interface for client-server communication
- Implemented the following functionalities: breakpoint, single-step, reading of registers and memory
- Implemented virtual-to-physical address translation, used by any guest memory access performed as part of the above-mentioned functionalities
- Created a new thread for handling client requests and two channels for vCPU - GDB server thread communication
- Added a "--debugger" command-line option for the firecracker executable

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The reason for this PR is clearly provided (issue no. or explanation).
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes are reflected in `firecracker/swagger.yaml`.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.

### CODE REVIEW

**Comment by gc-plp** [comment]
Hi @kumargu.

The gdb changes should be pointed towards the branch called `firecracker-microvm:feature/gdb_server`.

**Comment by kumargu** [comment]
> Hi @kumargu.
> 
> The gdb changes should be pointed towards the branch called `firecracker-microvm:feature/gdb_server`.

Yes, I will point it to the feature branch, just want to make sure everything builds fine with [this](https://github.com/firecracker-microvm/firecracker/pull/2319/) (https://github.com/firecracker-microvm/firecracker/pull/2319/) change which is not yet there in feature.

**Comment by dianpopa** [comment]
Hi @kumargu 

Wanted to know if you are still interested to contribute to this PR? Is there any help needed from us to get you unblocked?

**Comment by kumargu** [comment]
Hi,
Sorry, got a little busy with other stuff. 
Had a sync-up with @gc-plp. As next steps we will fix the build issues. The build is failing because we don't yet have support for aarch64. We will add flags to just support x86 for now and work for supporting aarch64 in a separate PR. 
I will raise a PR to fix the build issues and hopefully have a working build by this weekend.

**Comment by gc-plp** [comment]
Hi @kumargu. Seems that the commits should be signed. Please do that and resubmit.

**Comment by kumargu** [comment]
> Hi @kumargu. Seems that the commits should be signed. Please do that and resubmit.

Ack.

**Comment by kumargu** [comment]
Hi @dianpopa,

I noticed that you removed some dead code in this [commit](https://github.com/firecracker-microvm/firecracker/commit/e84ed0481bb0fbfaf778e97b79510426da6c8d1f#diff-2771b5bdabd1f363578ae32327132253aad690d29ffce7c0a7361b3a0876674d).

I just rebased this PR with master and noticed GDB server was using the util method, are you ok if I revert the commit in this PR? Other option would be to keep it within the gdb crate if it's not used outside.


Errors below  - 
```
 [18351] stderr:
E      Compiling kernel v0.1.0 (/firecracker/src/kernel)
E   error[E0433]: failed to resolve: could not find `structs` in `utils`
E      --> src/kernel/src/loader/mod.rs:292:16
E       |
E   292 |         utils::structs::read_struct_slice(kernel_image, ehdr.e_phnum as usize)
E       |                ^^^^^^^ could not find `structs` in `utils`
E   
E   error[E0425]: cannot find function `read_struct` in this scope
E      --> src/kernel/src/loader/mod.rs:283:9
E       |
E   283 |         read_struct(kernel_image, &mut ehdr)
E       |         ^^^^^^^^^^^ not found in this scope
E   
```

**Comment by dianpopa** [comment]
> commit

That commit you mentioned is part of this [PR](https://github.com/firecracker-microvm/firecracker/pull/2331) which replaces `read_struct_slice ` with ByteValued. My suggestion is to try to use ByteValued in your implementation too so as to keep consistency, WDYT?

**Comment by kumargu** [comment]
> > commit
> 
> That commit you mentioned is part of this [PR](https://github.com/firecracker-microvm/firecracker/pull/2331) which replaces `read_struct_slice ` with ByteValued. My suggestion is to try to use ByteValued in your implementation too so as to keep consistency, WDYT?

Agreed :100:

**Comment by kumargu** [comment]
@gc-plp this is ready for review.

File: src/gdb_server/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+pub use kernel::loader::elf::{Elf64_Phdr, PT_LOAD};
+pub use kvm_bindings;
+pub use kvm_ioctls::VcpuFd;
+pub use std::sync::mpsc::{Receiver, Sender};
+pub use vm_memory::{
+    ByteValued, Bytes, GuestAddress, GuestMemory, GuestMemoryMmap, GuestMemoryRegion,
+};
+
+pub use gdbstub::GdbStubError;
+#[allow(unused_imports)]
```
Comment (gc-plp) [inline src/gdb_server/src/lib.rs:19]: Can we remove `#[allow(unused_imports)]`? 

The purpose of imports is to have a clear understanding of what is used from other crates, but here we're just importing everything.

File: src/gdb_server/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+use gdbstub::{Connection, DisconnectReason, GdbStub, ResumeAction};
+
+extern crate vm_memory;
+
+#[allow(dead_code, unused_imports)]
```
Comment (gc-plp) [inline src/gdb_server/src/lib.rs:24]: Same for `allow(dead_code)`.

File: src/gdb_server/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+pub type DynResult<T> = Result<T, Box<dyn std::error::Error>>;
+
+#[allow(dead_code)]
+const PORT_NUM: u16 = 8443;
+
+#[allow(dead_code)]
+fn wait_for_tcp(port: u16) -> DynResult<TcpStream> {
+    let sockaddr = format!("0.0.0.0:{}", port);
+    eprintln!("Waiting for a GDB connection on {:?}...", sockaddr);
```
Comment (gc-plp) [inline src/gdb_server/src/lib.rs:41]: Instead of `eprintln` we could add these messages to the logger.

File: src/gdb_server/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    if target.insert_bp(entry_addr.0, false).is_err() {
+        return Err("GDB server error".into());
+    }
+
+    let connection: Box<dyn Connection<Error = std::io::Error>> =
+        { Box::new(wait_for_tcp(PORT_NUM)?) };
+    let mut debugger = GdbStub::new(connection);
+    match debugger.run(&mut target)? {
+        DisconnectReason::Disconnect => {
+            println!("Disconnected from GDB.");
```
Comment (gc-plp) [inline src/gdb_server/src/lib.rs:79]: `println` statements could also be directed through the logger.

File: src/gdb_server/src/target.rs [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+                val.2 = false;
+            }
+        }
+
+        Ok(())
+    }
+    pub fn insert_bp(&mut self, linear_addr: u64, translate: bool) -> Result<(), DebuggerError> {
+        // Opcode specific to x86 architecture that triggers a trap when
+        // encountered during cpu execution
+        let int3: u8 = 0xCC;
```
Comment (gc-plp) [inline src/gdb_server/src/target.rs:97]: I'd put this as a constant somewhere.

File: src/gdb_server/src/target.rs [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+        let mut opcode: Option<u8> = None;
+        if self.breakpoints_linear.contains_key(&linear_addr) {
+            let val = self.breakpoints_linear.get_mut(&linear_addr).unwrap();
+            if !val.2 {
```
Comment (gc-plp) [inline src/gdb_server/src/target.rs:101]: What do these tuples represent? Can we add a comment explaining the logic?

**Comment by kumargu** [comment]
sorry for the delay, will pick the comments this week.

**Comment by LegNeato** [comment]
Anything I can do to help here? I might have some free time to help push this over the line if you'll point me in the right direction.

**Comment by gc-plp** [comment]
Hi @LegNeato!

We are open to external contributors here as we don't have any work planned for this functionality.

The to-do items here would be:

1. Making page walking work. In order to translate from virtual to physical memory, the PoC in this PR makes use of a workaround where it picks up the sections from the elf file and uses that to do the translations. Ideally we want to avoid that.

2. Multi vcpu support: the PoC only works for 1-vcpu microvms, so we'd like to be able to use multi-vcpu microvms.

3. ARM support. Debugging arm microvms would be useful!

4. Hardware breakpoints support. The PoC only uses software breakpoints.

5. Enabling the gdb server code only when a special target is built - e.g. firecracker-gdb

The items above are not enumerated by priority. 

I'd say that (1), (5) and optionally (2) are must have to start working towards merging the PR.

If you're interested in working on any of these items I propose that we discuss here or on our slack server together with @kumargu.

Let us know what you think!

**Comment by kumargu** [comment]
Did a comparison of translated address using the current mechanism (page walks + reading phdr headers) vs multi level page walks. 
translated address matches for a) Identity mapped pages b)  1G/2M pages but the translation seems to fail for other cases (4K pages).

Some logs - 

```
Waiting for a GDB connection on "0.0.0.0:8443"...
Debugger connected from 127.0.0.1:59334
level 4 vaddr 10001f0 table-addr 9000 mask ffffffffff000 ent a023 offset 0
level 3 vaddr 10001f0 table-addr a023 mask ffffffffff000 ent b023 offset 0
level 2 vaddr 10001f0 table-addr b023 mask ffffffffff000 ent 10000a3 offset 40
Linear Address is: 16777712. Translated addresses using phdrs headers: 16777712 Translated Address using page walks: 16777712 Page size 2097152
level 4 vaddr ffffffff8100005d table-addr 9000 mask ffffffffff000 ent 0 offset ff8
level 3 vaddr ffffffff8100005d table-addr 0 mask ffffffffff000 ent 0 offset ff0
level 2 vaddr ffffffff8100005d table-addr 0 mask ffffffffff000 ent 0 offset 40
level 1 vaddr ffffffff8100005d table-addr 0 mask ffffffffff000 ent 0 offset 0
Linear Address is: 18446744071578845277. Translated addresses using phdrs headers: 16777309 Translated Address using page walks: 93 Page size 4096
level 4 vaddr ffffffff81ce5951 table-addr 9000 mask ffffffffff000 ent 0 offset ff8
level 3 vaddr ffffffff81ce5951 table-addr 0 mask ffffffffff000 ent 0 offset ff0
level 2 vaddr ffffffff81ce5951 table-addr 0 mask ffffffffff000 ent 0 offset 70
level 1 vaddr ffffffff81ce5951 table-addr 0 mask ffffffffff000 ent 0 offset 728
Linear Address is: 18446744071592368465. Translated addresses using phdrs headers: 30300497 Translated Address using page walks: 2385 Page size 4096
level 4 vaddr ffffffff81ce5966 table-addr 1d08000 mask ffffffffff000 ent 1c0c067 offset ff8
level 3 vaddr ffffffff81ce5966 table-addr 1c0c067 mask ffffffffff000 ent 1c0d063 offset ff0
level 2 vaddr ffffffff81ce5966 table-addr 1c0d063 mask ffffffffff000 ent 1c001e3 offset 70
Linear Address is: 18446744071592368486. Translated addresses using phdrs headers: 30300518 Translated Address using page walks: 30300518 Page size 2097152
level 4 vaddr ffffffff81ce5974 table-addr 1d08000 mask ffffffffff000 ent 1c0c067 offset ff8
level 3 vaddr ffffffff81ce5974 table-addr 1c0c067 mask ffffffffff000 ent 1c0d063 offset ff0
level 2 vaddr ffffffff81ce5974 table-addr 1c0d063 mask ffffffffff000 ent 1c001e3 offset 70
Linear Address is: 18446744071592368500. Translated addresses using phdrs headers: 30300532 Translated Address using page walks: 30300532 Page size 2097152
level 4 vaddr ffffffff81ce5982 table-addr 1d08000 mask ffffffffff000 ent 1c0c067 offset ff8
level 3 vaddr ffffffff81ce5982 table-addr 1c0c067 mask ffffffffff000 ent 1c0d063 offset ff0
level 2 vaddr ffffffff81ce5982 table-addr 1c0d063 mask ffffffffff000 ent 1c001e3 offset 70
Linear Address is: 18446744071592368514. Translated addresses using phdrs headers: 30300546 Translated Address using page walks: 30300546 Page size 2097152
level 4 vaddr ffffffff81ce5990 table-addr 1d08000 mask ffffffffff000 ent 1c0c067 offset ff8
level 3 vaddr ffffffff81ce5990 table-addr 1c0c067 mask ffffffffff000 ent 1c0d063 offset ff0
level 2 vaddr ffffffff81ce5990 table-addr 1c0d063 mask ffffffffff000 ent 1c001e3 offset 70
Linear Address is: 18446744071592368528. Translated addresses using phdrs headers: 30300560 Translated Address using page walks: 30300560 Page size 2097152
level 4 vaddr 3ff table-addr 1d08000 mask ffffffffff000 ent 0 offset 0
level 3 vaddr 3ff table-addr 0 mask ffffffffff000 ent 0 offset 0
level 2 vaddr 3ff table-addr 0 mask ffffffffff000 ent 0 offset 0
level 1 vaddr 3ff table-addr 0 mask ffffffffff000 ent 0 offset 0

```

Breakpoints - 
```
(gdb) target remote 127.0.0.1:8443
Remote debugging using 127.0.0.1:8443
0x0000000000000000 in irq_stack_union ()
(gdb) continue
Continuing.

Program received signal SIGTRAP, Trace/breakpoint trap.
0x0000000001000000 in ?? ()
(gdb) b *0x10001f0
Breakpoint 1 at 0x10001f0
(gdb) next
Cannot find bounds of current function
(gdb) b *0xffffffff8100005d
Breakpoint 2 at 0xffffffff8100005d
(gdb) b boot_cpu_init
Breakpoint 3 at 0xffffffff81ce5951
(gdb) continue
Continuing.

Breakpoint 1, 0x00000000010001f0 in ?? ()
(gdb) continue
Continuing.

Breakpoint 2, 0xffffffff8100005d in secondary_startup_64 ()
(gdb) continue
Continuing.

Breakpoint 3, 0xffffffff81ce5951 in boot_cpu_init ()
(gdb) next
Single stepping until exit from function boot_cpu_init,
which has no line number information.
0xffffffff81cd0b5d in start_kernel ()

```

**Comment by gc-plp** [comment]
Great to see this move forward @kumargu!

**Comment by kumargu** [comment]
I further did some more analysis 🧐 

translations through page walks fails only for addresses mapped to early bootup phase, because there is no notion of virtual address. To handle this baciumar@ had this nice logic to look at elf headers and adjust the addresses -- similar to what the linux kernel does [[link](https://github.com/torvalds/linux/blob/master/arch/x86/kernel/head64.c#L135)]. 

we would debate if we need this logic of parsing elf headers or just throw an exception if Firecracker GDB debuggers try to access memory regions in the early boot up phase.

**Comment by gc-plp** [comment]
@kumargu so page walking works after the mapping is done?

**Comment by kumargu** [comment]
@gc-plp Yes, it works after the mapping is done. there was a small check that was missing - I will update the PR.

**Comment by kumargu** [comment]
Update - 

* Address translation through page walks works once paging is enabled i,e post early boot-up process. 
* With current implementation, if someone puts a breakpoint within the memory regions mapped to early boot process, we will throw an exception. We will later work on solving this problem or use the logic to read elf headers -- if we see lot of people needs this feature.

Next steps - 

* Add unit tests to the page walking logic.
* Some code clean-ups.
* Build a different target for gdb -- essentially isolating it from firecracker's target build.

**Comment by dianpopa** [comment]
We will close this PR for the moment. @kumargu feel free to reopen if you have the bandwidth to resume work here.

**Pull request closed without merging**